File: Data Structures & Algorithms/count-subsequences/submission-5.cpp
class Solution {
public:
    vector<vector<int>> dp;
    int numDistinct(string s, string t) {
        int m = s.size(), n = t.size();
        if(n>m)
            return 0;
        dp = vector<vector<int>> (m+1, vector<int>(n+1, -1));
        return dfs(s, t, 0, 0);
    }

    int dfs(const string &s, const string &t, int i, int j){
        if(j == t.size())
            return 1;
        if(i == s.size())
            return 0;
        if(dp[i][j] != -1)
            return dp[i][j];
        int res;
        if(s[i] == t[j]){
            res = dfs(s, t, i+1, j+1) + dfs(s, t, i+1, j);
        }
        else{
            res = dfs(s, t, i+1, j);
        }
        dp[i][j] = res;
        return res;
    }
};

File: Data Structures & Algorithms/longest-common-subsequence/submission-5.py
class Solution:
    def longestCommonSubsequence(self, text1: str, text2: str) -> int:
        n = len(text1)
        m = len(text2)
        cache = {}

        def dfs(i,j):
            if (i,j) in cache:
                return cache[(i,j)]

            if i>=n or j>=m:
                return 0
            
            if text1[i]==text2[j]:
                ans =  1+dfs(i+1, j+1)
                cache[(i,j)] = ans
                return ans
            
            ans = max(dfs(i,j+1), dfs(i+1, j))
            cache[(i,j)] = ans
            return ans

        return dfs(0,0)